### PR TITLE
Add support for saving multiple metadata styles in the GUI

### PR DESCRIPTION
**Pull request by mizaki**
It may make more sense to put the `CheckableComboBox` in its own file in the `ui` dir?

I gave up with the style/theming and it seems a lot of hassle https://forum.qt.io/topic/113522/styling-qcombobox-s-lineedit

I presumed there is no point in allowing no write styles, so one is always required and the tooltip saying as much is displayed if it's tried. I thought this is probably the best place rather than say, checking on save etc.

A couple of places I changed reading of the current archive metadata from the write style to the read style. I can only think the idea was to keep what was already there? But if a read style is picked, then to my mind that should be used.

I'll do the read styles in another PR. With that, it may make sense to add another tab (or section) for the read and write styles in the GUI options.

### CODE REVIEW

File: comictaggerlib/pagelisteditor.py [hunk 1]
```diff
@@ -369,11 +369,12 @@ def emit_front_cover_change(self) -> None:
...
             self.firstFrontCoverChanged.emit(self.first_front_page)
 
-    def set_metadata_style(self, data_style: str) -> None:
+    def set_metadata_style(self, data_styles: list[str]) -> None:
         # depending on the current data style, certain fields are disabled
-        if data_style:
-            self.metadata_style = data_style
+        if data_styles:
+            styles = [metadata_styles[style] for style in data_styles]
+            enabled_widgets = [attr for style in styles for attr in style.supported_attributes]
```
Comment (lordwelch) [inline comictaggerlib/pagelisteditor.py]: This needs to be put into a proper for loop. Yes it's faster but it takes me longer than 10 seconds to figure out what is happening on this line

**Comment by lordwelch** [comment]
I'll test this out tomorrow.... probably

**Comment by mizaki** [comment]
Also, with the possible (seems like mostly dark style), when the read style is the same `CheckableComboBox` the look will be more unified as it kind of works if the read/write style widget look the same and the source one is a little different.

**Comment by mizaki** [comment]
Also, also is the question of converting over settings. As it's alpha should users be expected to remove/change the save style in the `settings.json`? Otherwise, it's leaving in code? Maybe change it over programatically and mark the code for removal at a later date?

File: comictaggerlib/taggerwindow.py [hunk 2]
```diff
@@ -1211,14 +1237,15 @@ def update_metadata_credit_colors(self) -> None:
 
     def update_metadata_style_tweaks(self) -> None:
         # depending on the current data style, certain fields are disabled
-
-        enabled_widgets = metadata_styles[self.save_data_style].supported_attributes
+        enabled_widgets = [
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Same as before write out the for loop here.

File: comictaggerlib/taggerwindow.py [hunk 3]
```diff
@@ -1791,36 +1833,40 @@ def identify_and_tag_single_archive(
...
+                        self.auto_tag_log(f"{metadata_styles[style].name()} save failed ;-(\n")
+                    else:
+                        match_results.good_matches.append(
+                            Result(
+                                Action.save,
+                                Status.success,
+                                ca.path,
+                                online_results=matches,
+                                match_status=MatchStatus.good_match,
+                            )
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: This should break the loop. Operating on files that are potentially corrupt can cause further data loss

File: comictaggerlib/taggerwindow.py [hunk 4]
```diff
@@ -1791,36 +1833,40 @@ def identify_and_tag_single_archive(
...
                         )
-                    )
-                    success = True
-                    self.auto_tag_log("Save complete!\n")
-                ca.load_cache([self.load_data_style, self.save_data_style])
+                        success = True
+                        self.auto_tag_log(f"{metadata_styles[style].name()} save complete!\n")
+
+                for style in self.save_data_styles:
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Why is this in a loop? The function takes a list.

File: comictaggerlib/ui/taggerwindow.ui [hunk 5]
```diff
@@ -59,30 +59,13 @@
...
-             </widget>
-            </item>
-            <item row="2" column="1">
-             <widget class="QComboBox" name="cbSaveDataStyle"/>
-            </item>
             <item row="0" column="0">
              <widget class="QLabel" name="lbl_md_source">
               <property name="text">
-               <string>Metadata Source</string>
+               <string>Data Source</string>
```
Comment (lordwelch) [inline comictaggerlib/ui/taggerwindow.ui:68]: Why was this renamed?

File: comictaggerlib/taggerwindow.py [hunk 6]
```diff
@@ -214,16 +214,31 @@ def __init__(
...
+        self.save_data_styles = config[0].internal__save_data_style
         self.load_data_style = config[0].internal__load_data_style
 
+        # Add multiselect combobox
+        self.cbSaveDataStyle = CheckableComboBox()
+        self.cbSaveDataStyle.setToolTip("At least 1 save style is required")
+        # Add normal combobox for read style (TODO support multiple read styles)
+        self.cbLoadDataStyle = QtWidgets.QComboBox()
+
+        # Need to set minimum or formLayout will resize larger than 230px (info and cover width)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: What is "info and cover width"

File: comictaggerlib/taggerwindow.py [hunk 7]
```diff
@@ -214,16 +214,31 @@ def __init__(
...
+        self.cbLoadDataStyle.setMinimumWidth(100)
+        self.cbSaveDataStyle.setMinimumWidth(100)
+        self.formLayout.addRow("Read Style", self.cbLoadDataStyle)
+        self.formLayout.addRow("Modify Styles", self.cbSaveDataStyle)
+        # Gets overwritten from ui file after using addRow
+        self.formLayout.setLabelAlignment(QtCore.Qt.AlignRight)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Give formLayout an actual name.

File: comictaggerlib/taggerwindow.py [hunk 8]
```diff
@@ -1460,7 +1488,8 @@ def populate_combo_boxes(self) -> None:
         self.cbFormat.addItem("Year One")
 
     def remove_auto(self) -> None:
-        self.remove_tags(self.save_data_style)
+        for style in self.save_data_styles:
+            self.remove_tags(style)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: remove_tags needs to take a list of styles instead of doing this. This will cause several popups to happen for every style selected

File: comictaggerlib/taggerwindow.py [hunk 9]
```diff
@@ -214,16 +214,31 @@ def __init__(
...
+        self.cbSaveDataStyle.setToolTip("At least 1 save style is required")
+        # Add normal combobox for read style (TODO support multiple read styles)
+        self.cbLoadDataStyle = QtWidgets.QComboBox()
+
+        # Need to set minimum or formLayout will resize larger than 230px (info and cover width)
+        self.cbLoadDataStyle.setMinimumWidth(100)
+        self.cbSaveDataStyle.setMinimumWidth(100)
+        self.formLayout.addRow("Read Style", self.cbLoadDataStyle)
+        self.formLayout.addRow("Modify Styles", self.cbSaveDataStyle)
+        # Gets overwritten from ui file after using addRow
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: How does this get overwritten from the ui file? It is loaded before this on line 89

**Review by lordwelch — CHANGES_REQUESTED** [review]
The checkable combobox is a decent idea. But currently clicking on the text does not consistently bring up the list on macos

File: comictaggerlib/ui/taggerwindow.ui (hunk 5 above)
Comment (mizaki) [reply comictaggerlib/ui/taggerwindow.ui:68]: I renamed it because it takes up too much space and means the more important names of the source or styles get more truncated.

File: comictaggerlib/taggerwindow.py (hunk 6 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: That is the information box and cover image under the source and styles form. I'll add a bit more detail to the message.

File: comictaggerlib/taggerwindow.py (hunk 4 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: Good question. I think I was thinking it was a list comprehension...

File: comictaggerlib/taggerwindow.py (hunk 9 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: It turns out even though I had saved with AlignRight (as it was before), because it was the default/unchanged it didn't save it and then defaulted to AlignLeft (which appears to still be incorrect as KDE should be AlignRight by default in the docs as I read it...). I cycled it between AlignLeft and AlignRight and then it saved.

**Comment by mizaki** [comment]
> The checkable combobox is a decent idea. But currently clicking on the text does not consistently bring up the list on macos

For me on Linux with KDE, I can spam click the text and it works as expected. Does it just highlight the text?
spam_click_write_style

**Comment by lordwelch** [comment]
They did some questionable things, which I've now fixed.
This should work better across multiple platforms, and at least on macOS it would previously change the style which it now does not
```python
    # Multiselect combobox from: https://gis.stackexchange.com/a/351152 (with custom changes)
    class CheckableComboBox(QtWidgets.QComboBox):
        itemChecked = pyqtSignal(str, bool)

        def __init__(self, *args: Any, **kwargs: Any):
            super().__init__(*args, **kwargs)
            # Prevent popup from closing when clicking on an item
            self.view().viewport().installEventFilter(self)

            # Keeps track of when the combobox list is shown
            self.justShown = False

        def resizeEvent(self, event: Any) -> None:
            # Recompute text to elide as needed
            super().resizeEvent(event)
            self._updateText()

        def eventFilter(self, obj: Any, event: Any) -> bool:
            # Allow events befor the combobox list is shown
            if obj == self.view().viewport():
                # We record that the combobox list has been shown
                if event.type() == QEvent.Show:
                    self.justShown = True
                # We record that the combobox list has hidden,
                # this will happen if the user does not make a selection
                # but clicks outside of the combobox list or presses escape
                if event.type() == QEvent.Hide:
                    self._updateText()
                    self.justShown = False
                # QEvent.MouseButtonPress is inconsistent on activation because double clicks are a thing
                if event.type() == QEvent.MouseButtonRelease:
                    # If self.justShown is true it means that they clicked on the combobox to change the checked items
                    # This is standard behavior (at least on macos) but I think it is surprising when it has a multiple select
                    if self.justShown:
                        self.justShown = False
                        return True

                    # Find the current index and item
                    index = self.view().indexAt(event.pos())
                    self.toggleItem(index.row())
                    return True
            return False

        def currentData(self) -> list[Any]:
            # Return the list of all checked items data
            res = []
            for i in range(self.count()):
                item = self.model().item(i)
                if item.checkState() == Qt.Checked:
                    res.append(self.itemData(i))
            return res

        def addItem(self, text, data=None):
            super().addItem(text, data)
            if self.count() == 1:
                self.model().item(0).setCheckState(Qt.Checked)

        def onDataChanged(self, index: Any) -> None:
            item = self.model().item(index.row())
            # If we have at least one item checked emit itemChecked with the current check state and update text
            # Require at least one item to be checked and provide a tooltip

        def _updateText(self) -> None:
            texts = []
            for i in range(self.count()):
                item = self.model().item(i)
                if item.checkState() == Qt.Checked:
                    texts.append(item.text())
            text = ", ".join(texts)

            # Compute elided text (with "...")

            # The QStyleOptionComboBox is needed for the call to subControlRect
            so = QtWidgets.QStyleOptionComboBox()
            # init with the current widget
            so.initFrom(self)

            # Ask the style for the size of the text field
            rect = self.style().subControlRect(QtWidgets.QStyle.CC_ComboBox, so, QtWidgets.QStyle.SC_ComboBoxEditField)

            # Compute the elided text
            elidedText = self.fontMetrics().elidedText(text, Qt.ElideRight, rect.width())

            # This CheckableComboBox does not use the index, so we clear it and set the placeholder text
            self.setCurrentIndex(-1)
            self.setPlaceholderText(elidedText)

        def setItemChecked(self, index: Any, state: bool) -> None:
            qt_state = Qt.Checked if state else Qt.Unchecked
            item = self.model().item(index)
            current = self.currentData()
            # Require at least one selection
            if len(current) == 1 and not state and item.checkState() == Qt.Checked:
                QtWidgets.QToolTip.showText(QtGui.QCursor.pos(), self.toolTip(), self, QRect(), 3000)
                return

            if len(current) > 0:
                item.setCheckState(qt_state)
                self.itemChecked.emit(self.itemData(index), state)
                self._updateText()

        def toggleItem(self, index: int) -> None:
            if self.model().item(index).checkState() == Qt.Checked:
                self.setItemChecked(index, Qt.Unchecked)
            else:
                self.setItemChecked(index, Qt.Checked)
```

https://github.com/comictagger/comictagger/assets/7547075/616d7b95-fefb-493d-907a-1d8fc49a17ec


https://github.com/comictagger/comictagger/assets/7547075/e272121b-3e4e-450e-8e4f-b63086758d54

**Comment by mizaki** [comment]
I believe that should be everything covered now?

I don't know why the hash has changed for the Cory cover image with the test.

File: comictaggerlib/pagelisteditor.py [hunk 10]
```diff
@@ -370,11 +370,16 @@ def emit_front_cover_change(self) -> None:
...
+    def set_metadata_style(self, data_styles: list[str]) -> None:
         # depending on the current data style, certain fields are disabled
-        if data_style:
-            self.metadata_style = data_style
+        if data_styles:
+            styles = [metadata_styles[style] for style in data_styles]
+            enabled_widgets = []
+            for style in styles:
+                for attr in style.supported_attributes:
+                    enabled_widgets.append(attr)
```
Comment (lordwelch) [inline comictaggerlib/pagelisteditor.py]: Lets change enabled_widgets to a set and you can can remove the inner loop:
```suggestion
        if data_styles:
            styles = [metadata_styles[style] for style in data_styles]
            enabled_widgets = set()
            for style in styles:
                enabled_widgets.update(style.supported_attributes)
```

File: comictaggerlib/taggerwindow.py [hunk 11]
```diff
@@ -1155,12 +1170,22 @@ def commit_metadata(self) -> None:
                 QtWidgets.QApplication.setOverrideCursor(QtGui.QCursor(QtCore.Qt.CursorShape.WaitCursor))
                 self.form_to_metadata()
 
-                success = self.comic_archive.write_metadata(self.metadata, self.save_data_style)
+                fail_list = []
+                # Save each style
+                for style in self.save_data_styles:
+                    success = self.comic_archive.write_metadata(self.metadata, style)
+                    if not success:
+                        fail_list.append(style)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: This should break, I'm not sure what happened you had fixed this earlier

File: comictaggerlib/autotagmatchwindow.py [hunk 12]
```diff
@@ -247,10 +247,10 @@ def save_match(self) -> None:
...
         md.overlay(ct_md)
-        success = ca.write_metadata(md, self._style)
-        ca.load_cache(list(metadata_styles))
-
-        QtWidgets.QApplication.restoreOverrideCursor()
+        for style in self._styles:
+            success = ca.write_metadata(md, style)
+            QtWidgets.QApplication.restoreOverrideCursor()
+            if not success:
+                QtWidgets.QMessageBox.warning(self, "Write Error", "Saving the tags to the archive seemed to fail!")
```
Comment (lordwelch) [inline comictaggerlib/autotagmatchwindow.py]: This needs to break on failure

File: comictaggerlib/taggerwindow.py [hunk 13]
```diff
@@ -1186,23 +1211,28 @@ def set_load_data_style(self, s: str) -> None:
...
 
     def update_metadata_credit_colors(self) -> None:
-        style = metadata_styles[self.save_data_style]
-        enabled = style.supported_attributes
+        styles = [metadata_styles[style] for style in self.save_data_styles]
+        enabled = []
+
+        for style in styles:
+            for attr in style.supported_attributes:
+                enabled.append(attr)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Same as above, turn this into a set and remove the inner loop

File: comictaggerlib/taggerwindow.py [hunk 14]
```diff
@@ -1212,14 +1242,17 @@ def update_metadata_credit_colors(self) -> None:
 
     def update_metadata_style_tweaks(self) -> None:
         # depending on the current data style, certain fields are disabled
+        enabled_widgets = []
+        for style in self.save_data_styles:
+            for attr in metadata_styles[style].supported_attributes:
+                enabled_widgets.append(attr)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Same as above, turn this into a set and remove the inner loop

File: comictaggerlib/taggerwindow.py [hunk 15]
```diff
@@ -1508,20 +1552,23 @@ def remove_tags(self, style: str) -> None:
...
                     QtCore.QCoreApplication.processEvents()
-                    if ca.has_metadata(style) and ca.is_writable():
-                        if not ca.remove_metadata(style):
-                            failed_list.append(ca.path)
-                        else:
-                            success_count += 1
-                        ca.load_cache(list(metadata_styles))
+                    for style in styles:
+                        if ca.has_metadata(style) and ca.is_writable():
+                            if not ca.remove_metadata(style):
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Lets invert the check here, negatives almost always make things more difficult to understand.

File: comictaggerlib/taggerwindow.py [hunk 16]
```diff
@@ -1508,20 +1552,23 @@ def remove_tags(self, style: str) -> None:
...
+                                failed_list.append(ca.path)
+                                # Abandon any further tag removals to prevent any greater damage to archive
+                                break
+                            else:
+                                success_count += 1
+                            ca.load_cache(list(metadata_styles))
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: lets also reset the cache before loading the cache

File: comictaggerlib/taggerwindow.py [hunk 17]
```diff
@@ -1538,9 +1585,13 @@ def copy_tags(self) -> None:
         has_src_count = 0
 
         src_style = self.load_data_style
-        dest_style = self.save_data_style
+        dest_styles = list(self.save_data_styles)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Is there a reason for the list call here?

File: comictaggerlib/taggerwindow.py [hunk 18]
```diff
@@ -1580,26 +1634,32 @@ def copy_tags(self) -> None:
...
+                            if prog_dialog.wasCanceled():
+                                break
+
+                            prog_dialog.setValue(prog_idx)
+                            prog_dialog.setLabelText(str(ca.path))
+                            center_window_on_parent(prog_dialog)
+                            QtCore.QCoreApplication.processEvents()
+
+                        if ca.has_metadata(src_style) and ca.is_writable():
+                            md = ca.read_metadata(src_style)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: This needs to be moved out of the loop, no reason to read the same metadata several times

File: comictaggerlib/taggerwindow.py [hunk 19]
```diff
@@ -1580,26 +1634,32 @@ def copy_tags(self) -> None:
...
-                        if not ca.write_metadata(md, dest_style):
-                            failed_list.append(ca.path)
-                        else:
-                            success_count += 1
+                            if not ca.write_metadata(md, style):
+                                failed_list.append(ca.path)
+                            else:
+                                if not ca_saved:
+                                    success_count += 1
+                                    ca_saved = True
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: You have an extra indent here

File: comictaggerlib/taggerwindow.py [hunk 20]
```diff
@@ -1580,26 +1634,32 @@ def copy_tags(self) -> None:
...
 
-                        ca.load_cache([self.load_data_style, self.save_data_style, src_style, dest_style])
+                        ca.load_cache([self.load_data_style, self.save_data_style, src_style, style])
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: this needs a de-dent, reset the cache before loading and the cache only needs to include `self.load_data_style` and `*self.save_data_styles`

File: comictaggerlib/taggerwindow.py [hunk 21]
```diff
@@ -1792,36 +1852,42 @@ def identify_and_tag_single_archive(
...
+                for style in self.save_data_styles:
+                    if not ca.write_metadata(md, style):
+                        match_results.write_failures.append(
+                            Result(
+                                Action.save,
+                                Status.write_failure,
+                                ca.path,
+                                online_results=matches,
+                                match_status=MatchStatus.good_match,
+                            )
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: There should be one result per comic archive and use these attributes, just like how the cli does
```python
    tags_deleted: list[str] = dataclasses.field(default_factory=list)
    tags_written: list[str] = dataclasses.field(default_factory=list)
```

File: comictaggerlib/taggerwindow.py [hunk 22]
```diff
@@ -1792,36 +1852,42 @@ def identify_and_tag_single_archive(
...
+                                Status.success,
+                                ca.path,
+                                online_results=matches,
+                                match_status=MatchStatus.good_match,
+                            )
+                        )
+                        success = True
+                        self.auto_tag_log(f"{metadata_styles[style].name()} save complete!\n")
+
+                ca.load_cache([self.load_data_style] + self.save_data_styles)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py:1882]: Lets reset the cache before loading the cache

File: comictaggerlib/taggerwindow.py [hunk 23]
```diff
@@ -1792,36 +1852,42 @@ def identify_and_tag_single_archive(
...
-                    match_results.write_failures.append(
-                        Result(
-                            Action.save,
-                            Status.write_failure,
-                            ca.path,
-                            online_results=matches,
-                            match_status=MatchStatus.good_match,
+                # Save each style
+                for style in self.save_data_styles:
+                    if not ca.write_metadata(md, style):
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: Lets invert the match here as well

File: comictaggerlib/taggerwindow.py [hunk 24]
```diff
@@ -215,15 +216,29 @@ def __init__(
...
+        self.cbSaveDataStyle.setToolTip("At least 1 save style is required")
+        # Add normal combobox for read style (TODO support multiple read styles)
+        self.cbLoadDataStyle = QtWidgets.QComboBox()
+
+        # Need to set minimum or source_style_formLayout will resize larger than 230px which will affect the
+        # file info box and cover image width underneath
+        self.cbLoadDataStyle.setMinimumWidth(100)
+        self.cbSaveDataStyle.setMinimumWidth(100)
+        self.source_style_formLayout.addRow("Read Style", self.cbLoadDataStyle)
+        self.source_style_formLayout.addRow("Modify Styles", self.cbSaveDataStyle)
```
Comment (lordwelch) [inline comictaggerlib/taggerwindow.py]: We can actually put all of this into the ui file.
Right click on the widget and click on promote
Screenshot 2024-02-10 at 1 22 52 PM

File: comictaggerlib/ui/taggerwindow.ui [hunk 25]
```diff
@@ -52,37 +52,23 @@
         <item>
          <layout class="QVBoxLayout" name="verticalLayout">
           <item>
-           <layout class="QFormLayout" name="formLayout">
+           <layout class="QFormLayout" name="source_style_formLayout">
```
Comment (lordwelch) [inline comictaggerlib/ui/taggerwindow.ui]: The comment on taggerwindow lines +230 to +241 removes the need for this to be named but.
`style_layout` is a much more manageable name

**Review by lordwelch — CHANGES_REQUESTED** [review]
The hash change was my last commit, I'll fix the test.

File: comictaggerlib/taggerwindow.py (hunk 11 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: Me too. I must have lost it in the squish...

File: comictaggerlib/autotagmatchwindow.py (hunk 12 above)
Comment (mizaki) [reply comictaggerlib/autotagmatchwindow.py]: I'll add the style name too otherwise it's going to be too vague.

File: comictaggerlib/taggerwindow.py (hunk 17 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: No...

File: comictaggerlib/taggerwindow.py (hunk 16 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: Should the `reset_cache` and `load_cache` be outside the loop?

File: comictaggerlib/taggerwindow.py (hunk 21 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: I believe this is correct now? I didn't see where any `tags_deleted` would come from?

File: comictaggerlib/taggerwindow.py (hunk 24 above)
Comment (mizaki) [reply comictaggerlib/taggerwindow.py]: I thought it would require a C header. Should have looked into it further...

File: comictaggerlib/taggerwindow.py (hunk 16 above)
Comment (lordwelch) [reply comictaggerlib/taggerwindow.py]: Outside of the style loop

File: comictaggerlib/taggerwindow.py (hunk 18 above)
Comment (lordwelch) [reply comictaggerlib/taggerwindow.py]: You also need to call continue when this check fails otherwise md may not exist when it tries to copy the tags

**Review by lordwelch — CHANGES_REQUESTED** [review]
The last thing I want to see is an option to use the ID instead of the full name for the checkable combo box so that I can actually see more than one at a time. It should be a file only option

**Comment by mizaki** [comment]
> The last thing I want to see is an option to use the ID instead of the full name for the checkable combo box so that I can actually see more than one at a time. It should be a file only option

Under the `General` tab? With the multi-read style I think a "Metadata Style" tab might be needed so I could add that now with that option?

**Comment by mizaki** [comment]
> > The last thing I want to see is an option to use the ID instead of the full name for the checkable combo box so that I can actually see more than one at a time. It should be a file only option
> 
> Under the `General` tab? With the multi-read style I think a "Metadata Style" tab might be needed so I could add that now with that option?

`Metadata Types` tab could be split up with a separator and the styles placed in there? I think a `listwidget` with the styles and checkboxes, so same as the combobox but in a list instead of a dropdown.

**Comment by mizaki** [comment]
> The last thing I want to see is an option to use the ID instead of the full name for the checkable combo box so that I can actually see more than one at a time. It should be a file only option

I set it to `False` by default but did consider `True` however showing the "friendly" name first made more sense.

When the read multi-styles is done I'll add the option to change the metadata styles in the settings window too.

The new function `populate_style_names` can be used for the read style when it goes in.

I'm not sure if there is a better way to go about updating the style names when the option has been checked?

**Review by lordwelch — APPROVED** [review]
